File: dotnet/src/MyDotey.CircularBuffer/CounterBucket.cs
using System;
using System.Collections.Concurrent;
using System.Threading;

namespace MyDotey.Util
{
    /**
     * @author koqizhao
     *
     * Jul 25, 2018
     */
    public class CounterBucket<T> : TimeBucket
    {
        private ConcurrentDictionary<T, AtomicLong> _counters;

        public CounterBucket(long startTime, long ttl)
            : base(startTime, ttl)
        {
            _counters = new ConcurrentDictionary<T, AtomicLong>();
        }

        public long Get(T identity)
        {
            if (identity == null)
                throw new ArgumentNullException("identity is null");

            _counters.TryGetValue(identity, out AtomicLong counter);
            return counter == null ? 0 : counter.Get();
        }

        public void Increment(T identity)
        {
            if (identity == null)
                throw new ArgumentNullException("identity is null");

            AtomicLong counter = GetCounter(identity);
            counter.IncrementAndGet();
        }

        public void Decrement(T identity)
        {
            if (identity == null)
                throw new ArgumentNullException("identity is null");

            AtomicLong counter = GetCounter(identity);
            counter.DecrementAndGet();
        }

        private AtomicLong GetCounter(T identity)
        {
            return _counters.GetOrAdd(identity, k => new AtomicLong());
        }

        public override void Reset(long startTime)
        {
            base.Reset(startTime);
            _counters.Clear();
        }
    }
}

File: java/circular-buffer/src/main/java/org/mydotey/util/CounterBucket.java
package org.mydotey.util;

import java.util.Objects;
import java.util.concurrent.ConcurrentHashMap;
import java.util.concurrent.atomic.AtomicLong;
import java.util.function.Function;

/**
 * @author koqizhao
 *
 * Jul 25, 2018
 */
public class CounterBucket<T> extends TimeBucket {

    protected static final Function<Object, AtomicLong> COUNTER_CREATOR = k -> new AtomicLong();

    private final ConcurrentHashMap<T, AtomicLong> _counters;

    public CounterBucket(long startTime, long ttl) {
        super(startTime, ttl);
        _counters = new ConcurrentHashMap<T, AtomicLong>();
    }

    public long get(T identity) {
        Objects.requireNonNull(identity, "identity is null");
        AtomicLong counter = _counters.get(identity);
        return counter == null ? 0 : counter.get();
    }

    public void increment(T identity) {
        Objects.requireNonNull(identity, "identity is null");
        AtomicLong counter = getCounter(identity);
        counter.incrementAndGet();
    }

    public void decrement(T identity) {
        Objects.requireNonNull(identity, "identity is null");
        AtomicLong counter = getCounter(identity);
        counter.decrementAndGet();
    }

    private AtomicLong getCounter(T identity) {
        return _counters.computeIfAbsent(identity, COUNTER_CREATOR);
    }

    @Override
    public void reset(long startTime) {
        super.reset(startTime);
        _counters.clear();
    }

}
